### PR TITLE
Add docker remote image tag function

### PR DESCRIPTION
**Pull request by hong-kisung**

### CODE REVIEW

File: vars/dockerCmd.groovy [hunk 1]
```diff
@@ -139,7 +139,231 @@ private def pushWithUsernameAndPassword(config, command, logger) {
       ${logoutCommand}
     """
   }
-}
+}
+
+def remoteTag(ret) {
+  Logger logger = Logger.getLogger(this)
+  logger.info "stared remote image change tag and push"
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:146]: stared -> started ??

File: vars/dockerCmd.groovy [hunk 2]
```diff
@@ -139,7 +139,231 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
-}
+}
+
+def remoteTag(ret) {
+  Logger logger = Logger.getLogger(this)
+  logger.info "stared remote image change tag and push"
+
+  def config = getParam(ret)
+
+  if(retParamErrorCheck(config, logger)) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:150]: if phrase is not necessary.
just run retParamErrorCheck(config, logger) alone

File: vars/dockerCmd.groovy [hunk 3]
```diff
@@ -139,7 +139,231 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+  if(!config.from) {
+    logger.error("from: projectName/ImageName:tag values or from: tag value is required.")
+    throw new RetortException('RC211')
+  }
+
+  if(!config.to) {
+    logger.error("to: projectName/ImageName:tag values or to: tag value is required.")
+    throw new RetortException('RC212')
+  }
+  return true
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:180]: Unnecessary

File: vars/dockerCmd.groovy [hunk 4]
```diff
@@ -139,7 +139,231 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+}
+
+private def convertNewMap(config, logger) {
+  logger.info "stared convertNewMap"
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:184]: stared -> started?

File: vars/dockerCmd.groovy [hunk 5]
```diff
@@ -139,7 +139,230 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+  if(remotImgMap.registry) {
+    tokenUrl.append("${remotImgMap.registry}")
+    tokenUrl.append(TOKEN_PATH)
+  }
+
+  if(remotImgMap.fromImage) {
+    tokenUrl.append("${remotImgMap.fromImage}:pull,push")
+  }
+
+  echo tokenUrl.toString()
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:237]: debug log??

File: vars/dockerCmd.groovy [hunk 6]
```diff
@@ -139,7 +139,230 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+
+  def responseBody
+  try {
+    responseBody = httpRequest httpMode: 'GET',
+    authentication: 'HARBOR_CREDENTIALS',
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:242]: what if remotImgMap.registry was not harbor??

File: vars/dockerCmd.groovy (hunk 1 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy]: Thank you. I will correct it.

File: vars/dockerCmd.groovy (hunk 2 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy]: Thank you. I will correct it.

File: vars/dockerCmd.groovy (hunk 3 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy]: Thank you. I will correct it.

File: vars/dockerCmd.groovy (hunk 4 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy]: Thank you. I will correct it.

File: vars/dockerCmd.groovy (hunk 5 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy]: yes,  I will correct it.(logger.debug "~~")

File: vars/dockerCmd.groovy (hunk 6 above)
Comment (jigyoomin) [reply vars/dockerCmd.groovy:240]: not HARBOR_CREDENTIALS, may be remotImgMap.credentialsId

File: vars/dockerCmd.groovy [hunk 7]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+
+  if(200 != responseBody.status) {
+    logger.error "failed to get token in Image Registry. [Error status code : [${responseBody.status}], message : ${responseBody.conatent}] "
+    throw createException('RC206')
+
+  }
+
+  def token = jsonMap.token
+
+  logger.debug "Token : ${token}"
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:264]: must not print token in log...

File: vars/dockerCmd.groovy [hunk 8]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+def remoteTag(ret) {
+  Logger logger = Logger.getLogger(this)
+  logger.info "started remote image change tag and push"
+
+  def config = getParam(ret)
+
+  retParamErrorCheck(config, logger)
+  def remotImgMap = convertNewMap(config, logger)
+  def token = getToken(remotImgMap, logger)
+  if(token) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:153]: if token was null of empty, it seems better to throw exception in getToken func.
this if block looks useless..

File: vars/dockerCmd.groovy [hunk 9]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+  def token = jsonMap.token
+
+  logger.debug "Token : ${token}"
+  return token
+}
+
+private def getManifestPath(token, remotImgMap, logger) {
+  logger.info "started get manifestInfo in image registry with - [ registry: ${remotImgMap.registry}, FromImage : ${remotImgMap.fromImage}, tagFromVersion : ${remotImgMap.fromVersion} ]"
+
+  def manifestType = "/v2/" + remotImgMap.fromImage + "/manifests/"
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:271]: remoteImageMap.fromImage could be null. need exception handling

File: vars/dockerCmd.groovy [hunk 10]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+    if(metaData){
+      pushRemoteImg(token, metaData, remotImgMap, logger)
+    }
+  }
+}
+
+private retParamErrorCheck(config, logger) {
+  logger.info "started request param error check"
+
+  if (!config.registry) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:164]: if registry is null, use HARBOR_REGISTRY to default value

File: vars/dockerCmd.groovy [hunk 11]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+}
+
+private def getManifestPath(token, remotImgMap, logger) {
+  logger.info "started get manifestInfo in image registry with - [ registry: ${remotImgMap.registry}, FromImage : ${remotImgMap.fromImage}, tagFromVersion : ${remotImgMap.fromVersion} ]"
+
+  def manifestType = "/v2/" + remotImgMap.fromImage + "/manifests/"
+  def contentType = "application/vnd.docker.distribution.manifest.v2+json"
+  StringBuffer getManifestUrl = new StringBuffer("https://")
+
+  if(remotImgMap.registry) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:275]: what if remotImgMap.registry is null??

File: vars/dockerCmd.groovy [hunk 12]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+    getManifestUrl.append("${remotImgMap.registry}")
+  }
+
+  getManifestUrl.append(manifestType)
+
+  if(remotImgMap.fromVersion) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:281]: not necessary

File: vars/dockerCmd.groovy [hunk 13]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+    contentType: 'APPLICATION_JSON',
+    customHeaders: [[name: 'Authorization', value: "Bearer " + token],[name: 'Accept', value: contentType]],
+    url: getManifestUrl.toString(),
+    quiet: false,
+    validResponseCodes: '100:599'
+  }catch(Exception e) {
+    throw createException('RC207', e)
+  }
+
+  if(200 != responseBody.status) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:299]: need to throw exception finally... in this if block.. cause http status is not 200

File: vars/dockerCmd.groovy [hunk 14]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+  }
+
+  logger.debug "success to get manifest in Image Registry : ${responseBody.content}"
+  return responseBody.content
+}
+
+private void pushRemoteImg(token, metaData, remotImgMap, logger) {
+  logger.info "started get manifestInfo in image registry with - [registry: ${remotImgMap.registry}, toImage : ${remotImgMap.toImage}, tagToVersion : ${remotImgMap.toVersion} ]"
+
+  def manifestType = "/v2/" + remotImgMap.toImage + "/manifests/"
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:321]: toImage could be null

File: vars/dockerCmd.groovy [hunk 15]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+  def contentType = "application/vnd.docker.distribution.manifest.v2+json"
+  StringBuffer getManifestUrl = new StringBuffer("https://")
+
+  if(remotImgMap.registry) {
+    getManifestUrl.append("${remotImgMap.registry}")
+  }
+
+  getManifestUrl.append(manifestType)
+
+  if(remotImgMap.toVersion) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:331]: not necessary

File: vars/dockerCmd.groovy [hunk 16]
```diff
@@ -139,7 +139,228 @@ private def pushWithUsernameAndPassword(config, command, logger) {
...
+    url: getManifestUrl.toString(),
+    requestBody: metaData,
+    quiet: false,
+    validResponseCodes: '100:599'
+    echo responseBody.content
+  }catch(Exception e) {
+    throw createException('RC209', e)
+  }
+
+  if(201 != responseBody.status) {
```
Comment (jigyoomin) [inline vars/dockerCmd.groovy:348]: 201 is right??

Comment (jigyoomin) [reply vars/dockerCmd.groovy:348]: and if only 201 is right, must throw exception at the end of block..

File: vars/dockerCmd.groovy (hunk 6 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:240]: authentication: remotImgMap.credentialsId 로 수정하겠습니다.

File: vars/dockerCmd.groovy (hunk 7 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:264]: 로그에서 token값 삭재하겠습니다

File: vars/dockerCmd.groovy (hunk 8 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:153]: getToken{}에서 아래와 같이 추가하겠습니다.
 if(!token?.trim()) { //the token is null or empty
    throw createException('RC214')
  }

File: vars/dockerCmd.groovy (hunk 9 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:271]: getManifestPath 타기전에 retParamErrorCheck()에서 아래와 같이 null이거나 ""이면 에러를 리턴하도록 수정하겠습니다.
 if(!config.from?.trim()) {
    logger.error("from: projectName/ImageName:tag values or from: tag value is required.")
    throw createException('RC211')
  }

File: vars/dockerCmd.groovy (hunk 10 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:164]: 말씀해주신대로 env.DEFAULT_HARBOR_REGISTRY를 사용하여, retParamErrorCheck()에서 아래와 같이 수정하겠습니다.
  if (!config.registry?.trim()) {
    def defaultHarborRegistry = getDefaultHarborRegistry()
    if (!defaultHarborRegistry?.trim()) {
      logger.error("if you want to use 'harbor', must set default defaultHarborRegistry or yourRegistry  ex) env.DEFAULT_HARBOR_REGISTRY = yourHarborRegistry or registry: yourHarborRegistry")
      throw createException('RC213')
    }
    config.registry = defaultHarborRegistry;
  }

File: vars/dockerCmd.groovy (hunk 11 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:275]: getParamErrorCheck()에서 체크하므로 if문 제외하고 아래와 같이 수정하겠습니다
 getManifestUrl.append("${remotImgMap.registry}")
  getManifestUrl.append(manifestType)
  getManifestUrl.append("${remotImgMap.fromVersion}")

File: vars/dockerCmd.groovy (hunk 12 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:281]: if 문 제외하겠습니다.

File: vars/dockerCmd.groovy (hunk 13 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:299]: 아래와 같이 if마지막에 추가하겠습니다.
 if(200 != responseBody.status) {
    def jsonMap
    try {
      jsonMap = readJSON text: responseBody.content
    }catch (Exception e) {
      logger.error "Failed to get manifest in Image Registry. [Error status code : [${responseBody.status}], message : ${responseBody.content}] "
      throw createException('RC208')
    }

    if (jsonMap.errors) {
      logger.error "Failed to get manifest in Image Registry. [Error status code : [${responseBody.status}], message : ${jsonMap.errors.message}] "
      throw createException('RC208')
    }
    throw createException('RC208',e)
  }

File: vars/dockerCmd.groovy (hunk 14 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:321]: retParamErrorCheck() 에서 아래와 같이 체크하도록 수정하겠습니다.
  if(!config.to?.trim()) {
    logger.error("to: projectName/ImageName:tag values or to: tag value is required.")
    throw createException('RC212')
  }

File: vars/dockerCmd.groovy (hunk 15 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:331]: 해당 if문 제외하겠습니다

File: vars/dockerCmd.groovy (hunk 16 above)
Comment (hong-kisung) [reply vars/dockerCmd.groovy:348]: 응답성공시 아래와 같이 201이 응답되고 있으며, 
13:19:29 HttpMethod: PUT
13:19:29 URL: https://mcm-dev-registry.cloudzcp.io/v2/kshong2/spring-boot-cicd-demo/manifests/prd-6
13:19:29 Authorization: *****
13:19:29 Content-Type: application/vnd.docker.distribution.manifest.v2+json
13:19:29 Sending request to url: https://mcm-dev-registry.cloudzcp.io/v2/kshong2/spring-boot-cicd-demo/manifests/prd-6
13:19:29 Response Code: HTTP/1.1 201 Created
13:19:29 Success code from [100‥599]

말씀해주신대로, 블록 마지막에 아래와 같이 throw exception 추가하겠습니다.
if(201 != responseBody.status) {
    def jsonMap
    try {
      jsonMap = readJSON text: responseBody.content
    }catch (Exception e) {
      logger.error "failed to push the remote image to the Image Registry. [Error status code : [${responseBody.status}], message : ${responseBody.content}] "
      throw createException('RC210')
    }

    if (jsonMap.errors) {
      logger.error "failed to push the remote image to the Image Registry. [Error status code : [${responseBody.status}], message : ${jsonMap.errors.message}] "
      throw createException('RC210')
    }
    throw createException('RC210',e)
  }

**Comment by hong-kisung** [comment]
위 사항 수정 후 다시 request pull 요청드리겠습니다~

**Pull request closed without merging**